windows/installer.nsi

```diff
@@ -1,213 +1,387 @@
-Delete "$SMPROGRAMS\${APP_NAME}\${APP_NAME}.lnk"
-Delete "$SMPROGRAMS\${APP_NAME}\${APP_NAME} server.lnk"
-Delete "$SMPROGRAMS\${APP_NAME}\${UNINSTALL_EXE}.lnk"
-RMDIR  "$SMPROGRAMS\${APP_NAME}"
-
-Delete $INSTDIR\${UNINSTALL_EXE}
-Delete $INSTDIR\${APP_EXE}
-Delete $INSTDIR\jamulus-server-icon-2020.ico
-Delete $INSTDIR\Qt5Core.dll
-Delete $INSTDIR\Qt5Gui.dll
-Delete $INSTDIR\Qt5Widgets.dll
-Delete $INSTDIR\Qt5Network.dll
-Delete $INSTDIR\Qt5Svg.dll
-Delete $INSTDIR\Qt5Concurrent.dll
-Delete $INSTDIR\Qt5Xml.dll
-Delete $INSTDIR\COPYING
-Delete $INSTDIR\platforms\qwindows.dll
-RMDir  $INSTDIR\platforms
-Delete $INSTDIR\styles\qwindowsvistastyle.dll
-RMDir  $INSTDIR\styles
-Delete $INSTDIR\imageformats\qgif.dll
-Delete $INSTDIR\imageformats\qicns.dll
-Delete $INSTDIR\imageformats\qico.dll
-Delete $INSTDIR\imageformats\qjpeg.dll
-Delete $INSTDIR\imageformats\qsvg.dll
-Delete $INSTDIR\imageformats\qtga.dll
-Delete $INSTDIR\imageformats\qtiff.dll
-Delete $INSTDIR\imageformats\qwbmp.dll
-Delete $INSTDIR\imageformats\qwebp.dll
-RMDir  $INSTDIR\imageformats
-RMDir  $INSTDIR
+Function ExitASIOInstalled
+    ClearErrors
+    EnumRegKey $0 HKLM "SOFTWARE\ASIO" 0
+    IfErrors 0 SkipMessage
+         MessageBox MB_YESNO|MB_ICONEXCLAMATION "$(ASIO_EXIT_NO_DRIVER)" /sd IDNO IDYES SkipMessage
+             Abort
+    SkipMessage:
+FunctionEnd
+
+; Uninstaller
+!macro un.InstallFiles buildArch
+
+    !define prefix "${DEPLOY_PATH}\${buildArch}"
+    !tempfile files
+
+    ; Find target files
+    !system 'cmd.exe /v /c "for /r "${prefix}" %f in (*.*) do \
+        @(set "_f=%f" && echo Delete "$INSTDIR\!_f:${prefix}\=!" >> "${files}")"'
+
+    ; Find target folders in reverse order to ensure they can be deleted when empty
+    !system 'cmd.exe /v /c "for /f "usebackq" %d in \
+        (`dir /b /s /ad "${prefix}" ^| C:\Windows\System32\sort.exe /r`) do \
+        @(set "_d=%d" && echo RMDir "$INSTDIR\!_d:${prefix}\=!" >> "${files}")"'
+
+    ; Remove files and folders
+    !include "${files}"
+
+    Delete "$INSTDIR\COPYING"
+    Delete "$INSTDIR\servericon.ico"
+    Delete "$INSTDIR\${UNINSTALL_EXE}"
+    RMDir  "$INSTDIR"
+
+    ; Cleanup
+    !delfile "${files}"
+    !undef files
+    !undef prefix
+
+!macroend
+
+Section "un.Install"
+    ; Delete the main application
+    ${If} ${RunningX64}
+        !insertmacro un.InstallFiles x86_64
+    ${Else}
+        !insertmacro un.InstallFiles x86
+    ${EndIf}
+
+    ; Remove the Start Menu and desktop shortcuts
+    IfFileExists "$DESKTOP\${APP_NAME}.lnk" deleteshortcut skipshortcut
+    deleteshortcut:
+      Delete   "$DESKTOP\${APP_NAME}.lnk"
+      goto skipshortcut
+    skipshortcut:
+    RMDir /r "$SMPROGRAMS\${APP_NAME}"
+
+    ; There may be an auto run entry in the registry for the server, remove it
+    DeleteRegValue HKCU "${AUTORUN_KEY}" "${AUTORUN_NAME}"
+
+    ; Remove the remaining registry keys
+    DeleteRegKey HKLM "${APP_UNINSTALL_KEY}"
+    DeleteRegKey HKLM "${APP_INSTALL_KEY}"
```

Comment:
@pljones it should remove all the registry content (including the path) here since I run the uninstaller on every new install of Jamulus (in my opinion it's the cleanest way for a update).

We might write the APP_INSTALL_KEY and your choices to a temp file or variable and load it with the installer after the uninstaller is finished. What do you think?

Reply:
If you've uninstalled, you're not expected to remember anything.

When I say remember my choices, I mean from one update to the next update.  The installer should read and store the values itself, run the uninstaller (which will clean up the registry), the use the values it previously read as the defaults, where applicable.  If the install completes, those values will be saved for the next run.

Reply:
Sorry, I think my answer was not clear.

Currently on an update of Jamulus, the installer runs the uninstaller of the last version if it finds one. Therefore, also the registry entries will be deleted. My proposal would be something like this:

```
if (oldUninstaller.exists()) {
    oldChoices = registry.getChoicesOfLastInstall();
    oldUninstaller.remove(); // will remove the old Jamulus version including the keys we just got
    registry.setChoicesOfInstall(oldChoices); // but since we saved them, we can rewrite them
}
```

Reply:
Yes, something like that.  If the installer reads the registry first, it should still have the values it read, even if it runs the uninstaller.  So "oldUninstaller.remove()" can still run the uninstaller.  Your installer is still running, just waiting for the uninstaller to finish.

Reply:
Ok. Now the state of "create desktop icon" is remembered #845 Can you please have a look at these changes and comment on the PR if this is enough?